src/par_iter/mod.rs

```diff
@@ -674,11 +671,114 @@ pub trait IndexedParallelIterator: ExactParallelIterator {
     /// iterators are of unequal length, you only get the items they
     /// have in common.
     fn zip<ZIP_OP>(self, zip_op: ZIP_OP) -> ZipIter<Self, ZIP_OP::Iter>
-        where ZIP_OP: IntoParallelIterator, ZIP_OP::Iter: IndexedParallelIterator
+        where ZIP_OP: IntoParallelIterator,
+              ZIP_OP::Iter: IndexedParallelIterator
     {
         ZipIter::new(self, zip_op.into_par_iter())
     }
 
+    /// Lexicographically compares the elements of this `ParallelIterator` with those of
+    /// another.
+    fn cmp<I>(self, other: I) -> Ordering
+        where I: IntoParallelIterator<Item = Self::Item>,
+              I::Iter: IndexedParallelIterator,
+              Self::Item: Ord
+    {
+        self.zip(other.into_par_iter())
+            .map(|(x, y)| Ord::cmp(&x, &y))
+            .reduce(|| Ordering::Equal,
+                    |cmp_a, cmp_b| if cmp_a != Ordering::Equal {
+                        cmp_b
+                    } else {
+                        cmp_a
+                    })
+    }
+
+    /// Lexicographically compares the elements of this `ParallelIterator` with those of
+    /// another.
+    fn partial_cmp<I>(self, other: I) -> Option<Ordering>
+        where I: IntoParallelIterator,
+              I::Iter: IndexedParallelIterator,
+              Self::Item: PartialOrd<I::Item>
+    {
+        self.zip(other.into_par_iter())
+            .map(|(x, y)| PartialOrd::partial_cmp(&x, &y))
+            .reduce(|| Some(Ordering::Equal),
+                    |cmp_a, cmp_b| match (cmp_a, cmp_b) {
```

Comment:
This is also not quite right. I think you want something like this:

```
match cmp_a {
    Some(Ordering::Equal) => cmp_b,
    Some(_) => cmp_a,
    None => None,
}
```

The idea is that once we find something which *is* ordered (and not equal), it doesn't matter what comes after. But if we find something unordered, same is true. 

Here is an example demonstrating this: http://play.integer32.com/?gist=87daf4489fb2ca744e1ea99abac0f443&version=stable